src/Microsoft.Azure.WebJobs.Host.Storage/BlobStorageConcurrencyStatusRepository.cs

```diff
@@ -0,0 +1,110 @@
+using Microsoft.Extensions.Configuration;
+using Microsoft.Extensions.Logging;
+using Newtonsoft.Json;
+
+namespace Microsoft.Azure.WebJobs.Host
+{
+    internal class BlobStorageConcurrencyStatusRepository : IConcurrencyStatusRepository
+    {
+        private const string HostContainerName = "azure-webjobs-hosts";
+        private readonly IHostIdProvider _hostIdProvider;
+        private readonly IConfiguration _configuration;
+        private readonly ILogger _logger;
+        private CloudBlobContainer _blobContainer;
+
+        public BlobStorageConcurrencyStatusRepository(IConfiguration configuration, IHostIdProvider hostIdProvider, ILoggerFactory loggerFactory)
+        {
+            _configuration = configuration;
+            _hostIdProvider = hostIdProvider;
+            _logger = loggerFactory.CreateLogger(LogCategories.Scale);
+        }
+
+        internal CloudBlobContainer Container
+        {
+            get
+            {
+                if (_blobContainer == null)
+                {
+                    string storageConnectionString = _configuration.GetWebJobsConnectionString(ConnectionStringNames.Storage);
+                    _blobContainer = CreateBlobContainer(storageConnectionString);
+                }
+                return _blobContainer;
+            }
+        }
+
+        public async Task<ConcurrencyStatusSnapshot> ReadAsync()
+        {
+            string blobPath = await GetBlobPathAsync();
+            try
+            {
+                CloudBlockBlob blob = Container.GetBlockBlobReference(blobPath);
+                if (await blob.ExistsAsync())
+                {
+                    string content = await blob.DownloadTextAsync();
+                    if (!string.IsNullOrEmpty(content))
+                    {
+                        var result = JsonConvert.DeserializeObject<ConcurrencyStatusSnapshot>(content);
+                        return result;
+                    }
+                }
+            }
+            catch (Exception e)
+            {
+                _logger.LogError(e, $"Error reading snapshot blob {blobPath}");
+                throw e;
+            }
+
+            return null;
+        }
+
+        public async Task WriteAsync(ConcurrencyStatusSnapshot snapshot)
+        {
+            var content = JsonConvert.SerializeObject(snapshot);
+
+            string blobPath = await GetBlobPathAsync();
+            try
+            {
+                CloudBlockBlob blob = Container.GetBlockBlobReference(blobPath);
+
+                using (StreamWriter writer = new StreamWriter(await blob.OpenWriteAsync()))
+                {
+                    await writer.WriteAsync(content);
+                }
+            }
+            catch (Exception e)
+            {
+                _logger.LogError(e, $"Error writing snapshot blob {blobPath}");
+                throw e;
+            }
+        }
+
+        private CloudBlobContainer CreateBlobContainer(string connectionString)
```

Comment:
Naming nit for additional code clarity.
```suggestion
        private CloudBlobContainer CreateBlobContainerIfNotExists(string connectionString)
```

Reply:
Renamed and made async